lib/phoenix_live_view/colocated_js.ex

```diff
@@ -0,0 +1,337 @@
+        File.rm!(Path.join(module_folder, file))
+      end
+
+      Enum.map(data, fn {filename, config} ->
+        absolute_file_path = Path.join(module_folder, filename)
+        config = Map.put(config, :app, Path.basename(app_dir))
+        {absolute_file_path, config}
+      end)
+    else
+      _ ->
+        # either the module does not exist any more or
+        # does not have any colocated hooks / JS
+        File.rm_rf!(module_folder)
+        []
+    end
+  end
+
+  defp get_data(module) do
+    case Phoenix.Component.MacroComponent.get_data(module) do
+      :error ->
+        []
+
+      data ->
+        Map.take(data, [
+          Phoenix.LiveView.ColocatedHook,
+          Phoenix.LiveView.ColocatedJS
+        ])
+        |> Enum.flat_map(fn {_mod, entries} -> entries end)
+    end
+  end
+
+  defp write_new_manifests!(target_dir, files) do
+    files
+    |> Enum.group_by(fn {_filename, config} -> config.app end)
+    |> Enum.each(fn {application, entries} ->
+      target_dir = Path.join(target_dir, application)
+
+      entries
+      |> Enum.group_by(fn {_file, config} ->
+        config[:manifest] || "index.js"
+      end)
+      |> Enum.each(fn {manifest, entries} ->
+        write_manifest(manifest, entries, target_dir)
+      end)
+    end)
+  end
+
+  defp write_manifest(manifest, entries, target_dir) do
+    content =
+      entries
+      |> Enum.group_by(fn {_file, config} -> config[:key] || :default end)
+      |> Enum.reduce([empty_manifest()], fn group, acc ->
+        case group do
+          {:default, entries} ->
+            [
+              acc,
+              Enum.map(entries, fn {file, %{name: name}} ->
+                import_name = "js_" <> Base.encode32(name, case: :lower, padding: false)
+                escaped_name = Phoenix.HTML.javascript_escape(name)
+
+                ~s<\nimport #{import_name} from "./#{Path.relative_to(file, target_dir)}"; js["#{escaped_name}"] = #{import_name};>
+              end)
+            ]
+
+          {key, entries} ->
+            escaped_key = Phoenix.HTML.javascript_escape(key)
+
+            [
+              acc,
+              ~s<js["#{escaped_key}"] = {};>,
+              Enum.map(entries, fn {file, %{name: name}} ->
+                import_name = "js_" <> Base.encode32(name, case: :lower, padding: false)
+                escaped_name = Phoenix.HTML.javascript_escape(name)
+
+                ~s<\nimport #{import_name} from "./#{Path.relative_to(file, target_dir)}"; js["#{escaped_key}"]["#{escaped_name}"] = #{import_name};>
+              end)
+            ]
+        end
+      end)
+
+    File.write!(Path.join(target_dir, manifest), content)
+  end
+
+  defp empty_manifest do
+    """
+    const js = {};
+    export default js;
+    """
+  end
+
+  defp target_dir do
+    Application.get_env(:phoenix_live_view, Phoenix.LiveView.ColocatedJS, [])
+    |> Keyword.get(:target_directory, Path.join(Mix.Project.build_path(), "phoenix-colocated"))
```

Comment:
I am wondering if we make it an app configuration or a `Mix.Project.config`. Given this is all about how the Mix compiler task work, I would suggest customizing it within `def project` (and we would read it as `Mix.Project.config`).

Reply:
I adjusted it to use the mix project config, but still allow a global setting if needed, since people might want to have colocated JS from dependencies put into another folder